platform/api/azure/instance.go

```diff
@@ -0,0 +1,228 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azure
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"regexp"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/arm/compute"
+	"github.com/Azure/azure-sdk-for-go/arm/network"
+
+	"github.com/coreos/mantle/util"
+)
+
+type Machine struct {
+	ID               string
+	PublicIPAddress  string
+	PrivateIPAddress string
+	InterfaceName    string
+	PublicIPName     string
+}
+
+func (a *API) getVMParameters(name, userdata, sshkey, storageAccountURI string, ip *network.PublicIPAddress, nic *network.Interface) compute.VirtualMachine {
+	osProfile := compute.OSProfile{
+		AdminUsername: util.StrToPtr("core"),
+		ComputerName:  &name,
+		LinuxConfiguration: &compute.LinuxConfiguration{
+			SSH: &compute.SSHConfiguration{
+				PublicKeys: &[]compute.SSHPublicKey{
+					{
+						Path:    util.StrToPtr("/home/core/.ssh/authorized_keys"),
+						KeyData: &sshkey,
+					},
+				},
+			},
+		},
+	}
+	if userdata != "" {
+		ud := base64.StdEncoding.EncodeToString([]byte(userdata))
+		osProfile.CustomData = &ud
+	}
+	var imgRef *compute.ImageReference
+	if a.opts.DiskURI != "" {
+		imgRef = &compute.ImageReference{
+			ID: &a.opts.DiskURI,
+		}
+	} else {
+		imgRef = &compute.ImageReference{
+			Publisher: &a.opts.Publisher,
+			Offer:     &a.opts.Offer,
+			Sku:       &a.opts.Sku,
+			Version:   &a.opts.Version,
+		}
+	}
+	return compute.VirtualMachine{
+		Name:     &name,
+		Location: &a.opts.Location,
+		Tags: &map[string]*string{
+			"createdBy": util.StrToPtr("mantle"),
+		},
+		VirtualMachineProperties: &compute.VirtualMachineProperties{
+			HardwareProfile: &compute.HardwareProfile{
+				VMSize: compute.VirtualMachineSizeTypes(a.opts.Size),
+			},
+			StorageProfile: &compute.StorageProfile{
+				ImageReference: imgRef,
+				OsDisk: &compute.OSDisk{
+					CreateOption: compute.FromImage,
+				},
+			},
+			OsProfile: &osProfile,
+			NetworkProfile: &compute.NetworkProfile{
+				NetworkInterfaces: &[]compute.NetworkInterfaceReference{
+					{
+						ID: nic.ID,
+						NetworkInterfaceReferenceProperties: &compute.NetworkInterfaceReferenceProperties{
+							Primary: util.BoolToPtr(true),
+						},
+					},
+				},
+			},
+			DiagnosticsProfile: &compute.DiagnosticsProfile{
+				BootDiagnostics: &compute.BootDiagnostics{
+					Enabled:    util.BoolToPtr(true),
+					StorageURI: &storageAccountURI,
+				},
+			},
+		},
+	}
+}
+
+func (a *API) CreateInstance(name, userdata, sshkey, resourceGroup, storageAccount string) (*Machine, error) {
```

Comment:
Should this do cleanup for partial failures or is that handled by the caller?

Reply:
At the end of every cluster's lifecycle the entire resource group is deleted which will remove any resources used by the test.